repository/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java

```diff
@@ -2,175 +2,123 @@
+        List<Predicate> tagPredicates = new ArrayList<>();
 
-    /**
-     * Updates an existing Gift Certificate in the database.
-     *
-     * @param giftCertificate The Gift Certificate to be updated.
-     */
-    @Override
-    public void update(GiftCertificate giftCertificate) {
-        log.info("Updating row in gift certificate table...");
-        jdbcTemplate.update(UPDATE_GIFT_CERTIFICATE,
-                giftCertificate.getName(),
-                giftCertificate.getDescription(),
-                giftCertificate.getPrice(),
-                giftCertificate.getDuration(),
-                giftCertificate.getId());
-        Integer count = jdbcTemplate.queryForObject(FIND_GIFT_CERTIFICATE_TAG_BY_ID, Integer.class, giftCertificate.getId());
-        if (count != null && count > 0) {
-            jdbcTemplate.update(DELETE_ALL_GIFT_CERTIFICATE_TAGS, giftCertificate.getId());
+        // Filtering by tags
+        if (searchFilter.tags() != null && !searchFilter.tags().isEmpty()) {
+            Join<GiftCertificate, Tag> tagJoin = root.join("getTags()");
+            for (Tag tag : searchFilter.tags()) {
+                Predicate tagPredicate = criteriaBuilder.equal(tagJoin.get("name"), tag.getName());
+                tagPredicates.add(tagPredicate);
+            }
+            finalPredicate = criteriaBuilder.and(tagPredicates.toArray(new Predicate[0]));
         }
-    }
 
-    /**
-     * Inserts tags associated with a Gift Certificate into the database.
-     *
-     * @param giftCertificate The Gift Certificate to which tags will be associated.
-     */
-    @Override
-    public void insertTags(GiftCertificate giftCertificate) {
-        log.info("Adding new rows into gift certificate_tag table...");
-        giftCertificate.getTags().forEach(tag -> jdbcTemplate.update(INSERT_TAG_TO_GIFT_CERTIFICATE, giftCertificate.getId(), tag.getId()));
-        setAllTags(giftCertificate);
+        // Filtering by name and description
+        if (!searchFilter.name().isEmpty() || !searchFilter.description().isEmpty()) {
+            Predicate searchPredicate = criteriaBuilder.or(
+                    criteriaBuilder.like(root.get("name"), "%" + searchFilter.name() + "%"),
+                    criteriaBuilder.like(root.get("description"), "%" + searchFilter.description() + "%")
+            );
+            finalPredicate = criteriaBuilder.and(finalPredicate, searchPredicate);
+        }
+
+        // Sorting
+        if (!searchFilter.sortType().isEmpty()) {
+            Comparator<GiftCertificate> comparator = giftCertificateComparators.get(searchFilter.sortType());
+            if (comparator != null) {
+                List<GiftCertificate> certificates = entityManager.createQuery(criteriaQuery).getResultList();
+                certificates.removeIf(giftCertificate -> !giftCertificate.getTags().containsAll(searchFilter.tags()));
+                certificates.sort(searchFilter.isDescending() ? comparator.reversed() : comparator);
```

Comment:
It is interesting. Why did you decide to use a comparator for sorting even  if you could use SQL query to get results?